### PR TITLE
GitHub Issue NOAA-EMC/GSI#441. Add bias predictors for ABI 2km data

### LINKED ISSUE #441: Add ABI 2km bias predictors for regional LAM
This issue is to describe work funded by a JTTI project investigating bias predictors for L1b ABI all-sky brightness temperature data for the regional model. Changes included in this repository are initially limited to the file setuprad.f90. The predictors are based on the work by Jason Otkin documented in:
Otkin, J. A., Potthast, R., & Lawless, A. S. (2018). Nonlinear Bias Correction for Satellite Data Assimilation Using Taylor Series Polynomials, Monthly Weather Review, 146(1), 263-285. Retrieved Jul 20, 2022, from https://journals.ametsoc.org/view/journals/mwre/146/1/mwr-d-17-0171.1.xml
Research was conducted at UW-Madison, CIMSS on the platform S4.  Using the water vapor channels of GOES-16 ABI (8-10), 4 predictors were defined using an offline calculation to optimize the nonlinear fit for the case period of October 2020 for a SE USA grid.

### PR DESCRIPTION
**Pull request by snebuda**

### CODE REVIEW

**Comment by MichaelLueken** [comment]
Hi @snebuda.  There are currently five commits associated with this work.  It will need to be reduced to a single commit.  To do this, please use the following commands:

- git clone --recursive git@github.com:snebuda/GSI.git update
- cd update
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git stash
- git remote add upstream https://github.com/NOAA-EMC/GSI.git
- git remote update
- git rebase upstream/develop
- git stash pop
- git add src/gsi/setuprad.f90
- git commit --amend
- Please add `GitHub Issue NOAA-EMC/GSI#441.` to the front of the commit message.
- git push origin develop --force

These steps will reduce the number of commits from 5 to 1, while maintaining your changes.

Also, please let me know who you would like to have review your changes.

If you have any questions, please let me know.

**Comment by snebuda** [comment]
@MichaelLueken-NOAA I work from UW-Madison S4 machine. When I clone GSI, it fails to check out fix files. This error is an old problem:

Submodule 'fix' (gerrit:GSI-fix) registered for path 'fix'
Cloning into 'fix'...
ssh: Could not resolve hostname vlab.ncep.noaa.gov: Name or service not known
fatal: Could not read from remote repository.
Please make sure you have the correct access rights
and the repository exists.
Clone of 'gerrit:GSI-fix' into submodule path 'fix' failed

When I issue the command >git commit --amend I get this error message:
U	fix
error: 'commit' is not possible because you have unmerged files.
...
fatal: Exiting because of an unresolved conflict.

Will the commit -a work or cause a problem?
If Emily Liu is available, she would be a good choice for reviewer.
Thanks - Sharon

**Comment by MichaelLueken** [comment]
@snebuda In that case, please try using these steps to reduce the number of commits from 5 to 1:

- git clone git@github.com:snebuda/GSI.git update
- cd update
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git reset --soft HEAD~1
- git stash
- git remote add upstream https://github.com/NOAA-EMC/GSI.git
- git remote update
- git rebase upstream/develop
- git stash pop
- git add src/gsi/setuprad.f90
- git commit --amend
- Please add `GitHub Issue NOAA-EMC/GSI#441.` to the front of the commit message.
- git push origin develop --force

This way, you shouldn't need to worry about the fix directory.

Please let me know if you encounter issues again.

**Comment by emilyhcliu** [comment]
@MichaelLueken-NOAA Please add @ADCollard for review.  Thanks.

File: src/gsi/setuprad.f90
Comment (emilyhcliu) [inline src/gsi/setuprad.f90:1196]: @snebuda I see that you are using pred(2:4,I) for abi2m bias corrections.  
Originally, pred(4:5,i) are predictors for 2nd-order and 1st-order lapse rate correction (air-mass bias correction) respectively.   
You only want 1st-order air-mass bias correction, right?

If you want to keep the original air-mass bias correction, you can consider using pred(7,i) or pred(8,i) for your 3rd-order abi2m correction.  The pred(7:8,i) are reserved for SSMIS.  They are zeros for non-SSMIS sensors.

**Review by emilyhcliu — COMMENTED** [review]
The code changes

**Comment by MichaelLueken** [comment]
Hi @emilyhcliu.  I have added @ADCollard as a reviewer.

File: src/gsi/setuprad.f90
Comment (snebuda) [reply src/gsi/setuprad.f90:1196]: @emilyhcliu Yes, we did not want to include the lapse rate corrections. I can move the term in the 4th slot to 7 or 8 if that is more consistent with other sensors' treatment. Or I can move the 3rd and 4th terms to pred(7:8,i)?

Comment (emilyhcliu) [reply src/gsi/setuprad.f90:1196]: @snebuda No, you do not need to move the predictors at all since you do not want the air-mass bias correction.

However, you will need to make sure you turn off the air-mass bias correction terms off completely.  The predoctors pred(4,i) and pred(5,i) are reserved for air-mass bias correction.  You took pred(4,i) for abi2m correction.  You will need to set pred(5,i) to zero. 

```
if (abi2m .and. regional) pred(5,i) = zero
```

Otherwise, you will have a correction from the first-order lapse rate term.
You can checked the abias file from your experiment and see which bias corrections are activated.

Comment (snebuda) [reply src/gsi/setuprad.f90:1196]: @emilyhcliu I have pred zeroed out on line 1191 within the "if (abi2km .and. regional) then" block:
pred(:,i) = zero
This new code block is after the lines 1153 & 1154:
           pred(4,i)=tlap*tlap
           pred(5,i)=tlap

Comment (emilyhcliu) [reply src/gsi/setuprad.f90:1196]: > @emilyhcliu I have pred zeroed out on line 1191 within the "if (abi2km .and. regional) then" block: pred(:,i) = zero This new code block is after the lines 1153 & 1154: pred(4,i)=tlap*tlap pred(5,i)=tlap

I see.  It is right at the beginning at the section you added.  Got it.  Thanks.

File: src/gsi/setuprad.f90
Comment (MichaelLueken) [inline src/gsi/setuprad.f90]: Please add proper precision to real values (i.e., 233.5_r_kind, 241.7_r_kind, and 250.5_r_kind)

File: src/gsi/setuprad.f90
Comment (MichaelLueken) [inline src/gsi/setuprad.f90]: Please add proper precision to real values (i.e., 190.0_r_kind and 300.0_r_kind)

File: src/gsi/setuprad.f90
Comment (MichaelLueken) [inline src/gsi/setuprad.f90]: In GNU and debug mode, the 4denvar configurations all failed on line 486 with the following message:

```
forrtl: severe (408): fort: (2): Subscript #1 of the array ABI2KM_BC has value 4 which is greater than the upper bound of 3

Image              PC                Routine            Line        Source
gsi.x              0000000007509490  Unknown               Unknown  Unknown
gsi.x              0000000004EF1B53  rad_setup_mp_setu         486  setuprad.f90
gsi.x              0000000003BB9879  gsi_radoper_mp_se         100  gsi_radOper.F90
gsi.x              00000000022B3B1D  setuprhsall_              490  setuprhsall.f90
gsi.x              0000000003B1EF8E  glbsoi_                   323  glbsoi.f90
gsi.x              0000000000E39300  gsisub_                   200  gsisub.F90
gsi.x              0000000000429641  gsimod_mp_gsimain        2212  gsimod.F90
gsi.x              000000000041223B  MAIN__                    631  gsimain.f90
gsi.x              000000000041209E  Unknown               Unknown  Unknown
libc-2.17.so       00002AEB1EB62555  __libc_start_main     Unknown  Unknown
gsi.x              0000000000411FA9  Unknown               Unknown  Unknown
```

Would it be best to move lines 483 through 486 to line 1190 to ensure that these are only given values for 2km ABI data?  Where they are currently situated, all nchanl values (including those less than four) will pass through this section, causing potential issues.  I made this change in a local copy of your fork and the code runs through to completion following the relocation of these lines to within the abi2km and regional if statement.

**Review by MichaelLueken — CHANGES_REQUESTED** [review]
Hi @snebuda.

I have completed my review of your changes.  The code compiles without issue using Intel and GNU compilers.  The code also successfully compiled in debug mode with no new warnings.  The regression tests were submitted, but issues were encountered with both GNU and debug tests, specifically for 4denvar configurations.  The regression tests were resubmitted following moving lines 483 through 486 to the abi2km and regional  if statement.  These configurations are now running and passing successfully.

I have also completed my visual inspection of these changes.  Please add proper precision to real values (i.e., 233.5_r_kind, 241.7_r_kind, 250.5_r_kind, 190.0_r_kind, and 300.0_r_kind).  Once these changes have been made and @emilyhcliu has given official approval, these changes should be ready for the review committee.

To ensure that only a single commit message is maintained, please use the following commands after modifying src/gsi/setuprad.f90:

```
git add src/gsi/setuprad.f90
git commit --amend
Please do not alter the current commit message, just save and close out of the commit message window.
git push origin develop --force
```

If you have any questions or encounter any issues, please let me know.

File: src/gsi/setuprad.f90
Comment (snebuda) [reply src/gsi/setuprad.f90]: Yes - I will move the abi2km terms assigned to constant values into the abi2km block of code. Thanks

Comment (MichaelLueken) [reply src/gsi/setuprad.f90]: Thank you very much for addressing my concerns!

**Review by MichaelLueken — DISMISSED** [review]
Hi @snebuda.  Thank you very much for addressing all of my concerns!  I approve of these code changes.  Once @emilyhcliu or @ADCollard give approval, this work will be ready for the review committee / authoritative repository.

File: src/gsi/setuprad.f90
Comment (MichaelLueken) [inline src/gsi/setuprad.f90]: One last quick modification, please add proper precision to real value 1. on line 1193 (i.e., 1._r_kind).

**Review by MichaelLueken — COMMENTED** [review]
Hi @snebuda.  One more minor modification is requested on line 1193, adding _r_kind to the 1.

Please use:

- git add src/gsi/setuprad.f90
- git commit --amend
- Please do not alter the current commit message, just save and close out of the commit message window.
- git push origin develop --force

to push ensure only a single commit message is maintained.  Thanks again!

File: src/gsi/setuprad.f90
Comment (MichaelLueken) [reply src/gsi/setuprad.f90]: Thanks for addressing this final update from me, @snebuda!

**Review by emilyhcliu — APPROVED** [review]

**Pull request merged**
